### PR TITLE
MQTT inbound added

### PR DESCRIPTION
**Pull request by elilmatha**

### CODE REVIEW

File: components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/protocol/mqtt/AxisMqttException.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
+/*
```
Comment (malakasilva) [inline components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/protocol/mqtt/AxisMqttException.java]: Why is this called axis2?

File: components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/protocol/mqtt/MqttAsyncCallback.java [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+     *                  subscription
+     * @throws MqttException
+     */
+    public void subscribe(String topicName, int qos) throws Throwable {
+        // Use a state machine to decide which step to do next. State change
+        // occurs
+        // when a notification is received that an MQTT action has completed
+        while (state != FINISH) {
+            switch (state) {
+                case BEGIN:
```
Comment (malakasilva) [inline components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/protocol/mqtt/MqttAsyncCallback.java]: We discussed that we don't need to wait for connect method. What are we still waiting for that?

File: components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/protocol/mqtt/MqttConnectionFactory.java [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+    private String factoryName;
+    private int retryInterval = 10000;
+    private int retryCount = 3;
+    private Hashtable<String, String> parameters = new Hashtable<String, String>();
+
+    public MqttConnectionFactory(Properties passedInParameter) {
+
+        this.factoryName = passedInParameter.getProperty(MqttConstants.PARAM_MQTT_CONFAC);
+
+        try {
```
Comment (malakasilva) [inline components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/protocol/mqtt/MqttConnectionFactory.java]: When a invalid or no parameter specified it'll only print a log and proceed? Have you tested such scenarios.

File: components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/protocol/mqtt/MqttConnectionFactory.java [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+
+            }
+
+            if (mqttClient == null) {
+                log.error("Error while creating the MQTT client");
+            } else {
+                log.info("Successfully created to mqtt client");
+            }
+
+        }//TODO -retry the connection
```
Comment (malakasilva) [inline components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/protocol/mqtt/MqttConnectionFactory.java]: remove the todo

File: components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/protocol/mqtt/MqttConnectionFactory.java [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+        try {
+            if (dataStore != null) {
+                mqttClient = new MqttClient(mqttEndpointURL, uniqueClientId, dataStore);
+            } else {
+                mqttClient = new MqttClient(mqttEndpointURL, uniqueClientId);
+            }
+            log.info("Successfully created to mqtt client");
+        } catch (MqttException e1) {
+            int retryC = 0;
+            while ((mqttClient == null) && ((retryCount == -1) || (retryC < retryCount))) {
```
Comment (malakasilva) [inline components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/protocol/mqtt/MqttConnectionFactory.java]: We discussed that we need to put a thread sleep for each retry. Why is it missing?

File: components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/protocol/mqtt/MqttConnectionFactory.java [hunk 6]
```diff
@@ -0,0 +1,283 @@
...
+            if (dataStore != null) {
+                mqttClient = new MqttAsyncClient(mqttEndpointURL, uniqueClientId, dataStore);
+            } else {
+                mqttClient = new MqttAsyncClient(mqttEndpointURL, uniqueClientId);
+            }
+            log.info("Successfully created to mqtt client");
+        } catch (MqttException e1) {
+            int retryC = 0;
+            while ((mqttClient == null) && ((retryCount == -1) || (retryC < retryCount))) {
+                retryC++;
```
Comment (malakasilva) [inline components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/protocol/mqtt/MqttConnectionFactory.java]: Same comment as previously mentioned

File: components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/protocol/mqtt/MqttListener.java [hunk 7]
```diff
@@ -0,0 +1,272 @@
...
+        }
+
+    }
+
+    @Override
+    public void init() {
+
+        log.info("MQTT inbound endpoint: " + name + " initializing ...");
+
+        if (mqttBlockingSenderEnable)
```
Comment (malakasilva) [inline components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/protocol/mqtt/MqttListener.java]: put brackets to this

File: components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/protocol/mqtt/MqttListener.java [hunk 8]
```diff
@@ -0,0 +1,272 @@
...
+        }
+
+        try {
+            mqttClient.connect();
+            if (confac.getTopic() != null) {
+                mqttClient.subscribe(confac.getTopic());
+            }
+            log.info("Connected to the remote server.");
+        } catch (MqttException e) {
+            if (!mqttClient.isConnected()) {
```
Comment (malakasilva) [inline components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/protocol/mqtt/MqttListener.java]: Make this retry interval as a parameter and optional. eg: when not specified default to 1000ms

File: components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/protocol/mqtt/MqttListener.java [hunk 9]
```diff
@@ -0,0 +1,272 @@
...
+                int retryC = 0;
+                while ((retryC < retryCount)) {
```
Comment (malakasilva) [inline components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/protocol/mqtt/MqttListener.java]: Can we set retry interval and counts?

File: components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/protocol/mqtt/MqttSyncCallback.java [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+     * Handle losing connection with the server. Here we just print it to the test console.
+     *
+     * @param throwable Throwable connection lost
+     */
+
+    @Override
+    public void connectionLost(Throwable throwable) { //TODO retry do in a single clas and use it.
+
+        connect();
+        log.info("Client : " + mqttClient.getClientId());
```
Comment (malakasilva) [inline components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/protocol/mqtt/MqttSyncCallback.java]: Why so many logs

File: components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/protocol/mqtt/MqttSyncCallback.java [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+            // Error handling goes here...
+        }
+    }
+
+    private void connect() {
+        boolean tryConnecting = true;
+        while (tryConnecting) {
+            try {
+                MqttClient mqttClient = confac.getMqttClient();
+
```
Comment (malakasilva) [inline components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/protocol/mqtt/MqttSyncCallback.java]: Unnecessary spacing

File: components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/protocol/mqtt/MqttAsyncCallback.java [hunk 12]
```diff
@@ -0,0 +1,285 @@
...
+    }
+
+    /**
+     * Handle losing connection with the server. Here we just print it to the test console.
+     *
+     * @param throwable Throwable connection lost
+     */
+    public void connectionLost(Throwable throwable) {
+        log.error("Connection Lost - Client Disconnected");
+    }// TODO ok
```
Comment (ashalya) [inline components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/protocol/mqtt/MqttAsyncCallback.java]: Remove unwanted comments

File: components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/protocol/mqtt/MqttConnectionFactory.java [hunk 13]
```diff
@@ -0,0 +1,288 @@
...
+                        mqttClient = new MqttClient(mqttEndpointURL, uniqueClientId, dataStore);
+                    } else {
+                        mqttClient = new MqttClient(mqttEndpointURL, uniqueClientId);
+                    }
+                    log.info("Successfully created to mqtt client");
+                } catch (MqttException e) {
+                    log.error("Error while trying to retry", e);
+                }
+
+            }
```
Comment (ashalya) [inline components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/protocol/mqtt/MqttConnectionFactory.java]: Remove empty lines

File: components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/protocol/mqtt/MqttInjectHandler.java [hunk 14]
```diff
@@ -0,0 +1,162 @@
...
+            if (log.isDebugEnabled()) {
+                log.debug("Processed MQTT Message of Content-type : "
+                        + contentType);
+            }
+            MessageContext axis2MsgCtx = ((org.apache.synapse.core.axis2.Axis2MessageContext) msgCtx)
+                    .getAxis2MessageContext();
+            // Determine the message builder to use
+
+            Builder builder = null;
+            if (contentType == null) {
```
Comment (ashalya) [inline components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/protocol/mqtt/MqttInjectHandler.java]: Use String utils to check null values

**Pull request merged**